crates/auto-hash-map/src/map.rs

```diff
@@ -0,0 +1,276 @@
+        }
+    }
+
+    pub fn contains_key(&self, key: &K) -> bool {
+        match self {
+            AutoMap::List(list) => list.iter().any(|(k, _)| *k == *key),
+            AutoMap::Map(map) => map.contains_key(key),
+        }
+    }
+}
+
+impl<K, V, H> AutoMap<K, V, H> {
+    pub fn iter(&self) -> Iter<'_, K, V> {
+        match self {
+            AutoMap::List(list) => Iter::List(list.iter()),
+            AutoMap::Map(map) => Iter::Map(map.iter()),
+        }
+    }
+
+    pub fn into_iter(self) -> IntoIter<K, V> {
+        match self {
+            AutoMap::List(list) => IntoIter::List(list.into_iter()),
+            AutoMap::Map(map) => IntoIter::Map(map.into_iter()),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            AutoMap::List(list) => list.len(),
+            AutoMap::Map(map) => map.len(),
+        }
+    }
+}
+
+pub enum Iter<'a, K, V> {
+    List(std::slice::Iter<'a, (K, V)>),
+    Map(std::collections::hash_map::Iter<'a, K, V>),
+}
+
+impl<'a, K, V> Iterator for Iter<'a, K, V> {
+    type Item = (&'a K, &'a V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Iter::List(iter) => iter.next().map(|(k, v)| (k, v)),
+            Iter::Map(iter) => iter.next().map(|(k, v)| (k, v)),
+        }
+    }
+}
+
+pub enum IntoIter<K, V> {
+    List(std::vec::IntoIter<(K, V)>),
+    Map(std::collections::hash_map::IntoIter<K, V>),
+}
+
+impl<K, V> Iterator for IntoIter<K, V> {
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            IntoIter::List(iter) => iter.next(),
+            IntoIter::Map(iter) => iter.next(),
+        }
+    }
+}
+
+pub enum Entry<'a, K, V> {
+    Occupied(OccupiedEntry<'a, K, V>),
+    Vacant(VacantEntry<'a, K, V>),
+}
+
+pub enum OccupiedEntry<'a, K, V> {
+    List {
+        list: &'a mut Vec<(K, V)>,
+        index: usize,
+    },
+    Map(std::collections::hash_map::OccupiedEntry<'a, K, V>),
+}
+
+impl<'a, K, V> OccupiedEntry<'a, K, V> {
+    pub fn get_mut(&mut self) -> &mut V {
+        match self {
+            OccupiedEntry::List { list, index } => &mut list[*index].1,
+            OccupiedEntry::Map(e) => e.get_mut(),
+        }
+    }
+
+    pub fn remove(self) -> V {
+        match self {
+            OccupiedEntry::List { list, index } => list.swap_remove(index).1,
+            OccupiedEntry::Map(e) => e.remove(),
+        }
+    }
+}
+
+pub enum VacantEntry<'a, K, V> {
+    List { list: &'a mut Vec<(K, V)>, key: K },
+    Map(std::collections::hash_map::VacantEntry<'a, K, V>),
+}
+
+impl<'a, K, V> VacantEntry<'a, K, V> {
+    pub fn insert(self, value: V) -> &'a mut V {
+        match self {
+            VacantEntry::List { list, key } => {
+                list.push((key, value));
+                &mut list.last_mut().unwrap().1
+            }
+            VacantEntry::Map(entry) => entry.insert(value),
+        }
+    }
+}
```

Comment:
Comment all the public items here. You can link to the corresponding `HashMap` APIs directly.